sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -364,11 +364,24 @@ class Analyzer(
 
         val (resolvedOrdering, missing) = resolveAndFindMissing(ordering, a, groupingRelation)
 
-        if (missing.nonEmpty) {
+        val addForAlias = new ArrayBuffer[NamedExpression]()
+        val aliasedOrdering = resolvedOrdering.zipWithIndex.map {
+          case (o, i) => {
+            o transform {
+              case aggOrSub @ (_: AggregateExpression | _: Substring) =>
```

Comment:
It's another bug and not only affect `Substring`... I tried `SELECT b + 1, count(*) FROM orderByData GROUP BY b + 1 ORDER BY b + 1` and also failed. The key problem is we build the temp grouping relation by `grouping.collect { case ne: NamedExpression => ne.toAttribute }` which filter out all un-named expressions like `Add`, `Substring`, etc. I will work on it.
cc @marmbrus @rxin